src/ILLink.Shared/TypeSystemProxy/ParameterProxy.cs

```diff
@@ -0,0 +1,54 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+
+namespace ILLink.Shared.TypeSystemProxy
+{
+	internal partial struct ParameterProxy
+	{
+		public ParameterProxy (MethodProxy method, ParameterIndex index)
+		{
+			if ((int) index < 0 || (int) index >= method.GetParametersCount ())
+				throw new InvalidOperationException ($"Parameter of index {(int) index} does not exist on method {method.GetDisplayName ()} with {method.GetParametersCount ()}");
+			Method = method;
+			Index = index;
+		}
+
+		public MethodProxy Method { get; }
+
+		public ParameterIndex Index { get; }
+
+		/// <summary>
+		/// The index of the entry in the '.parameters' metadata section corresponding to thie parameter.
+		/// Maps to the index of the parameter in Cecil's MethodReference.Parameters or Roslyn's IMethodSymbol.Parameters
+		/// Throws if the parameter is the implicit 'this' parameter.
+		/// </summary>
+		public int MetadataIndex {
+			get {
+				if (Method.HasImplicitThis ()) {
+					if (IsImplicitThis)
+						throw new InvalidOperationException ("Cannot get metadata index of the implicit 'this' parameter");
+					return (int) Index - 1;
+				}
+				return (int) Index;
+			}
+		}
+
+		// C# doesn't have partial properties, but the following should be implemented in every project
+
+		// public partial ReferenceKind RefKind { get; }
+
+		public partial string GetDisplayName ();
+
+		public bool IsImplicitThis => Method.HasImplicitThis () && Index == (ParameterIndex) 0;
+
+		public partial bool IsTypeOf (string typeName);
+
+		public IEnumerable<string> GetDiagnosticArgumentsForAnnotationMismatch ()
```

Comment:
I would prefer we "hide" the fact that implicit this parameter doesn't have a "name" and instead uses the method's name in the message alone. That is the GetDisplayName would be expected to not be called on the implicit this (maybe it throws?) and here we would just use the methods name directly.

As-is both this method and the implementation of GetDisplayName have to know that implicit this returns the method's name - which is weird (if it should return something, then probably constant `this`).

Reply:
That makes sense, I'll update GetDisplayName to return "this" and used Method.GetDisplayName() for the IsImplicitThis case of GetDiagnosticArgumentsForAnnotationsMismatch()